MRG +1: Fix for zero case

Eric89GXL opened the pull request:
@jasmainak @christianbrodbeck does this fix look right to you? This helps deal with the case where the covariance matrix is degenerate (e.g., some channel has a zero covariance / is only DC).

jasmainak commented on mne_sandbox/preprocessing/_sns.py:
```diff
@@ -112,7 +115,11 @@ def fit(self, raw, verbose=None):
+            norm = np.zeros(len(eigval))
+            use_mask = eigval > 0
+            norm[use_mask] = 1. / np.sqrt(eigval[use_mask])
+            eigvec *= norm
```
looks reasonable. I suppose it's the equivalent of L38 [here](http://audition.ens.fr/adc/NoiseTools/src/NoiseTools/doc/NoiseTools/nt_sns0.html) ?

jasmainak commented on mne_sandbox/preprocessing/_sns.py:
```diff
@@ -85,13 +85,16 @@ def fit(self, raw, verbose=None):
+        data_norm = np.diag(data_cov).copy()
+        pos_mask = data_norm > 0
+        data_norm[pos_mask] = 1. / data_norm
+        data_norm[~pos_mask] = 0
```
why do you have to fix it twice?

Eric89GXL commented:
That, and/or line 25. BTW I suspect but haven't actually shown that the use
of eps on 25 can distort the list of nearest neighbors for channels with
covariance around or less than eps(1) (such as clean gradiometers), so I
suspect that it's not the right way to protect against DC channels.

Eric89GXL commented:
Mathematically it can happen in both places (actually if it happens the
first time it should also happen the second, too. Think about the
degenerate case of interest and hopefully you'll agree.)

On Jan 13, 2017 15:52, "Mainak Jas" <notifications@github.com> wrote:

> *@jasmainak* commented on this pull request.
> ------------------------------
>
> In mne_sandbox/preprocessing/_sns.py
> <https://github.com/mne-tools/mne-sandbox/pull/25#pullrequestreview-16649603>
> :
>
> >          # re-index this
>          good_picks = np.searchsorted(picks, good_picks)
>          bad_picks = np.setdiff1d(np.arange(len(picks)), good_picks)
>          data_cov[np.ix_(good_picks, good_picks)] = good_cov
>          del good_picks
> -        data_norm = np.diag(data_cov)
> +        data_norm = np.diag(data_cov).copy()
> +        pos_mask = data_norm > 0
> +        data_norm[pos_mask] = 1. / data_norm
> +        data_norm[~pos_mask] = 0
>
> why do you have to fix it twice?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-sandbox/pull/25#pullrequestreview-16649603>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACQZXkw_wVegBb2xM1Hpo2dl0GR7ozyvks5rR-QfgaJpZM4LjRQO>
> .
>

Eric89GXL commented:
> BTW I suspect but haven't actually shown that the use of eps on 25 can distort the list of nearest neighbors for channels with covariance around or less than eps(1)

Actually I guess it's pretty easy to show. Create `a` and `b` with correlation 1 but different scales, and `c` with correlation < 1 with `a` but closer scale:
```
>>> a = np.arange(-2, 3).astype(float)
>>> b = a * 1e-16
>>> c = [-3, -1, 0, 1, 3]
>>> data = np.array([a, b, c], float)
>>> cov = np.dot(data, data.T)
# This one correctly lists a and b as nearest neighbors
>>> cov / (np.sqrt(np.diag(cov)) * np.sqrt(np.diag(cov))[:, np.newaxis])
array([[ 1.        ,  1.        ,  0.98994949],
       [ 1.        ,  1.        ,  0.98994949],
       [ 0.98994949,  0.98994949,  1.        ]])
# this one does not
>>> eps = np.finfo(float).eps
>>> eps
2.2204460492503131e-16
>>> cov / (np.sqrt(np.diag(cov) + eps) * np.sqrt(np.diag(cov) + eps)[:, np.newaxis])
array([[  1.00000000e+00,   2.12216861e-08,   9.89949494e-01],
       [  2.12216861e-08,   4.50359963e-16,   2.10083975e-08],
       [  9.89949494e-01,   2.10083975e-08,   1.00000000e+00]])
```

jasmainak commented:
okay point taken about the `eps`. I think in Alain's code, you may not have this problem because L65 [here](http://audition.ens.fr/adc/NoiseTools/src/NoiseTools/doc/NoiseTools/nt_sns.html) removes high amplitude artifacts ...?
Sorry but I don't follow your solution yet. Do you propose to normalize the cov only for channels where the diag(cov) > 0 ?

jasmainak commented on mne_sandbox/preprocessing/_sns.py:
```diff
@@ -85,13 +85,16 @@ def fit(self, raw, verbose=None):
-        data_norm = np.diag(data_cov)
+        data_norm = np.diag(data_cov).copy()
+        pos_mask = data_norm > 0
+        data_norm[pos_mask] = 1. / data_norm
```
the shape of `pos_mask` and `data_norm` is not different?

Eric89GXL commented:
I think in Alain's code, you may not have this problem because L65 here
<http://audition.ens.fr/adc/NoiseTools/src/NoiseTools/doc/NoiseTools/nt_sns.html>
removes high amplitude artifacts ...?


This isn't about high amplitude artifacts but rather DC/zero values.

Sorry but I don't follow your solution yet. Do you propose to normalize the
cov only for channels where the diag(cov) > 0 ?

Yes, you cannot normalize those channels, you end up trying to normalize 0
by 0.

jasmainak commented:
> This isn't about high amplitude artifacts but rather DC/zero values.
I thought the epsilon example you gave would be problematic only if the scales were very different -> you have high amplitude artifacts, no?

Travis fails and moreover maybe you want to update a test?

Eric89GXL commented:
Yeah the fix has some bugs (like the one you pointed out), I should be able to fix them next week. Thanks for the look

Eric89GXL commented:
> I thought the epsilon example you gave would be problematic only if the scales were very different -> you have high amplitude artifacts, no?

If the scales of one channel versus another are quite different -- or really if they're anywhere near `eps(1)`, which gradiometers are.

Eric89GXL commented:
Okay, cleaned up the code, and added a similarity test to the output of NoiseTools. It now also reconstructs bad channels using good channels rather than just leaving them unmodified.

Eric89GXL commented:
Requires https://github.com/mne-tools/mne-python/pull/3909 to pass

Eric89GXL commented:
Okay I think this should be good to go now

jasmainak commented on mne_sandbox/preprocessing/_sns.py:
```diff
@@ -64,6 +73,11 @@ def fit(self, raw, verbose=None):
+        Notes
+        -----
+        In the resulting operator, bad channels will be reconstructed by
+        using the good channels.
```
why does SNS need to know bad channels? I think you mentioned it somewhere but I can't find the comment any more.

Eric89GXL replied:
Bad channels will be reconstructed using the good channels, but bad channels will not contribute to the good ones (i.e., the bad channel columns in the operator will be zero). This is the difference.

jasmainak replied:
okay so if `raw.info['bads']` is empty list, then it's the same as the NoiseTools implementation, and otherwise not?

Eric89GXL replied:
Yeah

jasmainak replied:
okay cool. You have my blessings to merge then :)

jasmainak commented on mne_sandbox/preprocessing/tests/test_sns.py:
```diff
@@ -86,5 +89,31 @@ def test_sns():
 
+    # Test against NoiseTools
+    rng = np.random.RandomState(0)
+    n_channels = 9
```
perhaps add a bad channel here by multiplying the data in one channel by some factor?

Eric89GXL replied:
PR welcome to improve later. Don't want to do it in this first pass since NoiseTools SNS doesn't reconstruct bad channels in the same way

Eric89GXL replied:
(and regarding the scaling issue that initiated this PR, SNS doesn't handle that explicitly so the comparison would break)

Eric89GXL commented:
Thanks for the reviews @jasmainak
